analysis/src/python/data_analysis/preprocessing/build_submissions_series.py

```diff
@@ -0,0 +1,113 @@
+import argparse
+import logging
+import sys
+
+import pandas as pd
+
+from analysis.src.python.data_analysis.model.column_name import SubmissionColumns
+from analysis.src.python.data_analysis.utils.df_utils import append_df, write_df
+from analysis.src.python.data_analysis.utils.parsing_utils import str_to_datetime
+
+
+def check_same_code(submission_0: pd.Series, submission_1: pd.Series) -> bool:
+    """ Check if submission not the same as previous. """
+
+    return submission_0[SubmissionColumns.CODE] == submission_1[SubmissionColumns.CODE]
+
+
+def check_different_code(submission_0: pd.Series, submission_1: pd.Series, diff_ratio: float) -> bool:
+    """ Check if submission not so different with previous (lines number does not changed in `diff_ratio` times). """
+
+    code_0 = submission_0[SubmissionColumns.CODE]
+    code_1 = submission_1[SubmissionColumns.CODE]
+    code_lines_diff = len(code_0) / len(code_1)
+    return not (1 / diff_ratio <= code_lines_diff <= diff_ratio)
+
+
+def filter_submissions_series(submissions_series: pd.DataFrame, diff_ratio: float) -> pd.DataFrame:
+    """ Filter submissions in submission series (group of submissions by one user on one step). """
+
+    logging.info(f'Processing submission series group {submissions_series.iloc[0][SubmissionColumns.GROUP]}')
+    logging.info(f'Initial group shape {submissions_series.shape}')
+    submissions_series = submissions_series.copy()
+    submissions_series[SubmissionColumns.TIME] = submissions_series[SubmissionColumns.TIME].apply(str_to_datetime)
+    submissions_series.sort_values([SubmissionColumns.TIME], inplace=True)
+
+    i = 1
+    while i < submissions_series.shape[0]:
+
+        submission_0 = submissions_series.iloc[i - 1]
+        submission_1 = submissions_series.iloc[i]
+        if check_same_code(submission_0, submission_1):
+            logging.info(f'Drop submission: '
+                         f'user={submission_1[SubmissionColumns.USER_ID]} '
+                         f'step={submission_1[SubmissionColumns.STEP_ID]} '
+                         f'attempt={i + 1}: '
+                         f'same submissions')
+            submissions_series.drop(submissions_series.iloc[i].name, inplace=True, axis=0)
+        elif check_different_code(submission_0, submission_1, diff_ratio):
+            logging.info(f'Drop submission: '
+                         f'user={submission_1[SubmissionColumns.USER_ID]} '
+                         f'step={submission_1[SubmissionColumns.STEP_ID]} '
+                         f'attempt={i + 1}: '
+                         f'different submissions')
+            submissions_series.drop(submissions_series.iloc[i].name, inplace=True, axis=0)
```

Comment:
It is not a very efficient way to iterate over two rows by one iteration. Consider the following approach:
1. Create a tmp column with a short, e.g. `df[SHIFTED_FRAGMENT] = df[FRAGMENT].shift(1)`
2. Use df.apply. In this step you have access to the current fragment and the next fragment

Also you can add a flag if the columns should be deleted or something like this, delete them after handling and delete tmp columns

It will be more readable and faster